### PR TITLE
cap filename length when attaching media to cards

### PR DESCRIPTION
**Pull request by Arthur-Milchior**
## Purpose / Description
Main purpose: having libanki similar to Anki's folder Anki.

## Fixes
Avoid a potential risk when media name added through android are extremly long.

Note that I believe that dae's code itself does not make sens here. Indeed, the same collection can be synchronized on any device. So the same constraints should apply to all devices. Otherwise, we risks to have media which can be used on android and not on windows. 

However, here, I'm only porting anki's code to ankidroid and thus don't intend to correct it

## Approach
As in Anki

## How Has This Been Tested?

gradlew. 
I fail to test this, because I fail to produce the bug it's supposed to avoid. When I try to use a virtual machine, I can't rename a file to give it more than 1024 characters. Mostly because typing a single character takes seconds. Which means that I can actually have a file name long enough to ensure that current ankidroid have a problem and that this droid solve it.

However I tested that we can still add normal files in ankidroid, and that it works perfectly.

## Checklist

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code

### CODE REVIEW

**Comment by mikehardy** [comment]
Here's the commit relying on an Android API26 API while we support down to API15

```

05:10:56 V/InstrumentationResultParser: INSTRUMENTATION_STATUS: stack=java.lang.NoClassDefFoundError: java.nio.file.Paths
05:10:56 V/InstrumentationResultParser: at com.ichi2.libanki.Media._cleanLongFilename(Media.java:575)
05:10:56 V/InstrumentationResultParser: at com.ichi2.libanki.Media.cleanFilename(Media.java:553)
05:10:56 V/InstrumentationResultParser: at com.ichi2.libanki.Media.writeData(Media.java:243)
05:10:56 V/InstrumentationResultParser: at com.ichi2.libanki.Media.addFile(Media.java:225)
05:10:56 V/InstrumentationResultParser: at com.ichi2.anki.tests.libanki.MediaTest.testAdd(MediaTest.java:70)
```

**Comment by mikehardy** [comment]
can't use Paths API in a project supporting API15

**Comment by Arthur-Milchior** [comment]
I removed paths in new commit.
I also edited text to explain how ideally someone would test this

**Comment by Arthur-Milchior** [comment]
Codacy remarks corrected (i.e. I use "".equals(fname) instead of fname.equals(""))

**Comment by Arthur-Milchior** [comment]
Can you please restart API 25 which is the only one which failed ?

**Comment by mikehardy** [comment]
Can you rebase this to master and re-push? There are quite a few travis changes aimed at hopefully reducing flakiness while increasing coverage that went into the .travis.yml in the last couple days

**Comment by Arthur-Milchior** [comment]
Done

File: AnkiDroid/src/main/java/com/ichi2/libanki/Media.java [hunk 1]
```diff
@@ -544,6 +546,54 @@ public boolean hasIllegal(String str) {
...
+            if (lastDot == -1 || lastDot < lastSlash) {
+                // no dot, or before last slash
+                fname = fname.substring(0, namemax);
+            } else {
+                String ext = fname.substring(lastDot+1);
+                String head = fname.substring(0, lastDot);
+                int headmax = namemax - ext.length();
+                head = head.substring(0, headmax);
+                fname = head + ext;
+                assert (fname.length() <= namemax);
```
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/libanki/Media.java]: Let's use `Assert()`

Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/libanki/Media.java]: Sorry for keeping forgetting that. It's done

Comment (mikehardy) [reply AnkiDroid/src/main/java/com/ichi2/libanki/Media.java]: I checked to see if I could make Codacy warn on these, and with checkstyle rules it does not seem possible to do a regex checks against method names, so I'll have just have to scan closely for that capital A vs lower A

https://github.com/checkstyle/checkstyle/blob/78ef8b45492302ac7bf2cd27a01ee07c674b136b/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/IllegalTokenTextCheck.java#L108-L119

**Comment by mikehardy** [comment]
@timrae - this is a collaboration question/verification. I had approved it earlier but missed the assert/Assert. You approved it now. At this point here (and future PRs this happens) I'd say I could merge and we'd be within current process?

**Comment by timrae** [comment]
@mikehardy
Yeah if I've approved something but haven't merged it's usually just because I'm waiting for the CI or something like that, so if it's all green and approved by me, and you also have checked the code and approved it then please feel free to merge

**Pull request merged**